test/test_array_data.cpp

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "doctest/doctest.h"
+
+#include <initializer_list>
+#include "sparrow/array_data.hpp"
+
+namespace sparrow
+{
+    class mock_layout
+    {
+    public:
+
+        using value_type = std::optional<int>;
+        using inner_value_type = value_type::value_type;
+        using reference = reference_proxy<mock_layout>;
+        using size_type = std::size_t;
+        using storage_type = std::vector<std::optional<int>>;
+
+        mock_layout() = default;
+        mock_layout(std::initializer_list<value_type> l)
+            : m_storage(l)
+        {
+        }
+
+        reference operator[](size_type pos)
+        {
+            return reference(*this, pos);
+        }
+
+        const storage_type& storage() const
+        {
+            return m_storage;
+        }
+
+    private:
+
+        bool has_value(size_type index) const
+        {
+            return m_storage[index].has_value();
+        }
+
+        typename value_type::value_type& value(size_type index)
```

Comment:
How about?
```suggestion
        typename inner_value_type& value(size_type index)
```

Reply:
Ah indeed, I forgot to replace this. The typename is not even required with the change.